packages/router-api/src/primitives.rs

```diff
@@ -245,6 +223,219 @@ impl AsRef<str> for ChainName {
     }
 }
 
+#[cw_serde]
+#[serde(try_from = "String")]
+#[derive(Eq, Hash)]
+pub struct LegacyChainName(String);
+
+impl FromStr for LegacyChainName {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.contains(CHAIN_NAME_DELIMITER) || s.is_empty() {
+            return Err(Error::InvalidChainName);
+        }
+
+        Ok(LegacyChainName(s.to_owned()))
+    }
+}
+
+impl From<LegacyChainName> for String {
+    fn from(d: LegacyChainName) -> Self {
+        d.0
+    }
+}
+
+impl TryFrom<String> for LegacyChainName {
+    type Error = Error;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        value.parse()
+    }
+}
+
+impl Display for LegacyChainName {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl AsRef<str> for LegacyChainName {
+    fn as_ref(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+impl<'a> PrimaryKey<'a> for LegacyChainName {
+    type Prefix = ();
+    type SubPrefix = ();
+    type Suffix = Self;
+    type SuperSuffix = Self;
+
+    fn key(&self) -> Vec<Key> {
+        vec![Key::Ref(self.0.as_bytes())]
+    }
+}
+
+impl<'a> Prefixer<'a> for LegacyChainName {
+    fn prefix(&self) -> Vec<Key> {
+        vec![Key::Ref(self.0.as_bytes())]
+    }
+}
+
+impl KeyDeserialize for LegacyChainName {
+    type Output = Self;
+
+    #[inline(always)]
+    fn from_vec(value: Vec<u8>) -> StdResult<Self::Output> {
+        String::from_utf8(value)
+            .map_err(StdError::invalid_utf8)?
+            .then(LegacyChainName::try_from)
+            .map_err(StdError::invalid_utf8)
+    }
+}
+
+impl KeyDeserialize for &LegacyChainName {
+    type Output = ChainName;
+
+    #[inline(always)]
+    fn from_vec(value: Vec<u8>) -> StdResult<Self::Output> {
+        ChainName::from_vec(value)
+    }
+}
+
+#[cw_serde]
+#[serde(try_from = "String", into = "String")]
+#[derive(Eq, Hash)]
+pub enum GeneralizedChainName {
+    Amplifier(ChainName),
+    Legacy(LegacyChainName),
+}
+
+impl GeneralizedChainName {
+    pub fn amplifier(self) -> Result<ChainName, Error> {
+        match self {
+            GeneralizedChainName::Amplifier(chain) => Ok(chain),
+            GeneralizedChainName::Legacy(_) => Err(Error::InvalidChainName),
+        }
+    }
+
+    pub fn legacy(self) -> Result<LegacyChainName, Error> {
+        match self {
+            GeneralizedChainName::Amplifier(_) => Err(Error::InvalidChainName),
```

Comment:
I feel like this could probably use it's own error, such as `InvalidChainType` or somesuch